cmd/zrepl/zrepl.c

```diff
@@ -546,114 +602,120 @@ get_controller_ip_address(char *buf, int len)
+	if (rc == -1) {
+		ZREPL_ERRLOG("epoll_ctl() failed with errno:%d\n", errno);
+		goto exit;
 	}
 
-	printf("Controller IP address is: %s\n", target_addr);
-	bzero((char *)&istgt_addr, sizeof (istgt_addr));
-	istgt_addr.sin_family = AF_INET;
-	istgt_addr.sin_addr.s_addr = inet_addr(target_addr);
-	istgt_addr.sin_port = htons(TARGET_PORT);
-retry:
-	rc = connect(sfd, (struct sockaddr *)&istgt_addr, sizeof (istgt_addr));
-	if ((rc == -1) && ((errno == EINTR) || (errno == ECONNREFUSED) ||
-	    (errno == ETIMEDOUT) || (errno == EINPROGRESS))) {
-		ZREPL_ERRLOG("Failed to connect to istgt_controller"
-		    " with err:%d\n", errno);
-		sleep(2);
-		printf("Retrying ....\n");
-		goto retry;
-	} else {
-		printf("Connection to TGT controller successful\n");
-		ZREPL_LOG("Connection to TGT controller iss successful\n");
-	}
+	/* Buffer where events are returned */
+	events = calloc(MAXEVENTS, sizeof (event));
 
 	while (1) {
-		rc = uzfs_zvol_read_header(sfd, &hdr);
-		if (rc < 0) {
+		int i, n = 0;
+		n = epoll_wait(efd, events, MAXEVENTS, 1000);
+		if (n  == 0) {
+			printf("Event triggered by timer\n");
+			printf("Value of n is %d\n", n);
+			uzfs_zinfo_update_io_seq_for_all_volumes();
+			continue;
+		}
+
+		for (i = 0; i < n; i++) {
+			if ((events[i].events & EPOLLRDHUP) ||
+			    (events[i].events & EPOLLHUP) ||
+			    (events[i].events & EPOLLERR)) {
+				ZREPL_ERRLOG("epoll err() :%d\n", errno);
 close_conn:
-			ZREPL_ERRLOG("Management connection disconnected\n");
-			/*
-			 * Error has occurred on this socket
-			 * close it and open a new socket after
-			 * 5 sec of sleep.
-			 */
-			close(sfd);
-			printf("Retrying ....\n");
-			sleep(5);
-			sfd = create_and_bind(mgmt_port, false);
-			if (sfd == -1)
-				goto exit;
+				close(events[i].data.fd);
+				sfd = uzfs_zvol_connect_to_tgt_controller(arg);
+				if (sfd == -1) {
+					goto exit;
+				}
+				event.data.fd = sfd;
+				event.events = flags;
+				rc = epoll_ctl(efd, EPOLL_CTL_ADD, sfd, &event);
+				if (rc == -1) {
+					ZREPL_ERRLOG("epoll_ctl() failed "
+					    "with errno:%d\n", errno);
+					goto exit;
+				}
+				continue;
+			}
 
-retry1:
-			rc = connect(sfd, (struct sockaddr *)&istgt_addr,
-			    sizeof (istgt_addr));
-			if ((rc == -1) && ((errno == EINTR) ||
-			    (errno == ECONNREFUSED) || (errno == ETIMEDOUT))) {
-				ZREPL_ERRLOG("Failed to connect to"
-				    " istgt_controller with err:%d\n",
-				    errno);
-				sleep(2);
-				goto retry1;
-			} else {
-				printf("Connection to TGT controller "
-				    "successful\n");
-				ZREPL_LOG("Connection to TGT controller"
-				    "is successful\n");
+			rc = uzfs_zvol_read_header(events[i].data.fd,
+			    &hdr, O_NONBLOCK);
+			if (rc < 0) {
+				ZREPL_ERRLOG("Management connection "
+				    "disconnected\n");
+				/*
+				 * Error has occurred on this socket
+				 * close it and open a new socket after
+				 * 5 sec of sleep.
+				 */
+				goto close_conn;
+			} else if (rc > 0) {
```

Comment:
unit test case to hit this

Reply:
I think Jan has already have UT case to cover this branch.

Reply:
yes, this is covered in C++ tests.